packages/@aws-cdk/core/lib/validation/report.ts

```diff
@@ -0,0 +1,186 @@
+   */
+  readonly locations: string[];
+
+  /**
+   * The path to the CloudFormation template that contains this resource
+   */
+  readonly templatePath: string;
+}
+
+/**
+ * Construct violating a specific rule.
+ */
+export interface ValidationViolatingConstruct extends ValidationViolatingResource {
+  /**
+   * The construct path as defined in the application.
+   */
+  readonly constructPath: string;
+
+  /**
+   * A stack of constructs that lead to the violation.
+   *
+   * @default - stack will be empty if the cli is not run with `--debug`
+   */
+  readonly constructStack?: string;
+}
+
+/**
+ * Validation produced by the validation plugin, in construct terms.
+ */
+export interface ValidationViolationConstructAware extends ValidationViolation {
+  /**
+   * The constructs violating this rule.
+   */
+  readonly violatingConstructs: ValidationViolatingConstruct[];
+}
+
+/**
+ * Validation produced by the validation plugin, in CFN resource terms
+ */
+export interface ValidationViolationResourceAware extends ValidationViolation {
+  /**
+   * The resources violating this rule.
+   */
+  readonly violatingResources: ValidationViolatingResource[];
+}
+
+
+/**
+ * The final status of the validation report
+ */
+export enum ValidationReportStatus {
+  /**
+   * No violations were found
+   */
+  SUCCESS = 'success',
+
+  /**
+   * At least one violation was found
+   */
+  FAILURE = 'failure',
+}
+
+/**
+ * Summary of the report.
+ */
+export interface ValidationReportSummary {
+  /**
+   * The final status of the validation (pass/fail)
+   */
+  readonly status: ValidationReportStatus;
+
+  /**
+   * The name of the plugin that created the report
+   */
+  readonly pluginName: string;
+
+  /**
+   * Additional metadata about the report. This property is intended
+   * to be used by plugins to add additional information.
+   *
+   * @default - no metadata
+   */
+  readonly metadata?: { readonly [key: string]: string };
+}
+
+/**
+ * JSON representation of the report.
+ */
+export interface ValidationReportJson {
+  /**
+   * Report title.
+   */
+  readonly title: string;
+
+  /**
+   * Reports for all of the validation plugins registered
+   * in the app
+   */
+  readonly pluginReports: PluginReportJson[];
+}
+
+/**
+ * A report from a single plugin
+ */
+export interface PluginReportJson {
+  /**
+   * List of violations in the report.
+   */
+  readonly violations: ValidationViolationConstructAware[];
+
+  /**
+   * Report summary.
+   */
+  readonly summary: ValidationReportSummary;
+}
+
+/**
+ * The report emitted by the plugin after evaluation.
+ */
+export interface ValidationReport {
+  /**
+   * List of violations in the report.
+   */
+  readonly violations: ValidationViolationResourceAware[];
+
+  /**
+   * Whether or not the report was successful.
+   */
+  readonly success: boolean;
+
+  /**
+   * The name of the plugin that created the report
+   */
+  readonly pluginName: string;
```

Comment:
This shouldn't be necessary right? the framework can just call `plugin.name` to get this.

Reply:
That's true. But as it is, the same interface is used by the plugin and by the formatter. We'd have to split it into a base report that doesn't have the plugin name and another interface that extends the base one, adding the plugin name. The base would be used by the plugin and the extended one, by the framework. Is that what you're thinking?

Reply:
Was hoping to avoid having to extend. The formatter can be passed the plugin name when you call `formatString` - no? And then it just incorporates it into what it produces.

Reply:
Oh, so the idea is to keep the `pluginName` property, and just make it optional?

Reply:
Yeah I think the extension solution is the less convoluted. I'm going ahead with it.